src/borg/cache.py

```diff
@@ -389,6 +371,203 @@ def rollback(self):
+            # we see in the filesystem.
+            self.files[path_hash] = msgpack.packb(entry._replace(inode=st.st_ino, age=0))
+            return entry.chunk_ids
+        else:
+            return None
+
+    def memorize_file(self, path_hash, st, ids):
+        if not stat.S_ISREG(st.st_mode):
+            return
+        entry = FileCacheEntry(age=0, inode=st.st_ino, size=st.st_size, mtime=st.st_mtime_ns, chunk_ids=ids)
+        self.files[path_hash] = msgpack.packb(entry)
+        self._newest_mtime = max(self._newest_mtime or 0, st.st_mtime_ns)
+
+    def commit(self):
+        if self.files is None:
+            return
+        if self._newest_mtime is None:
+            # was never set because no files were modified/added
+            self._newest_mtime = 2 ** 63 - 1  # nanoseconds, good until y2262
+        ttl = int(os.environ.get('BORG_FILES_CACHE_TTL', 20))
+        with SaveFile(os.path.join(self.cache.path, 'files'), binary=True) as fd:
+            for path_hash, item in self.files.items():
+                # Only keep files seen in this backup that are older than newest mtime seen in this backup -
+                # this is to avoid issues with filesystem snapshots and mtime granularity.
+                # Also keep files from older backups that have not reached BORG_FILES_CACHE_TTL yet.
+                entry = FileCacheEntry(*msgpack.unpackb(item))
+                if entry.age == 0 and entry.mtime < self._newest_mtime or \
+                                      entry.age > 0 and entry.age < ttl:
+                    msgpack.pack((path_hash, entry), fd)
+
+    def _read_files(self):
+        self.files = {}
+        self._newest_mtime = None
+        logger.debug('Reading files cache ...')
+        with open(os.path.join(self.cache.path, 'files'), 'rb') as fd:
+            u = msgpack.Unpacker(use_list=True)
+            while True:
+                data = fd.read(64 * 1024)
+                if not data:
+                    break
+                u.feed(data)
+                for path_hash, item in u:
+                    entry = FileCacheEntry(*item)
+                    # in the end, this takes about 240 Bytes per file
+                    self.files[path_hash] = msgpack.packb(entry._replace(age=entry.age + 1))
+
+
+class ChunksCache:
+    def __init__(self, cache):
+        self.cache = cache
+
+    def add_chunk(self, id, chunk, stats, overwrite=False):
+        """
+        Add *chunk* with *id* to the underlying repository, updating *stats*.
+
+        If *overwrite* is true, then an existing chunk is replaced, if any;
+        otherwise the data in *chunk* is ignored if a chunk with *id* already
+        exists.
+
+        Return a ChunkListEntry.
+        """
+
+
+    def seen_chunk(self, id, size=None):
+        """
+        Return whether a chunk with *id* was seen. Optionally verify *size* for
+        enhanced collision resistance.
+        """
```

Comment:
Currently, this seems to either return a boolean (seen / not seen) or an int (refcount, >0 = seen, 0 = not seen).
Can we avoid the type confusion?